Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp

```diff
@@ -46,9 +50,34 @@ WebPermissionControllerProxy::~WebPermissionControllerProxy()
     m_process.removeMessageReceiver(Messages::WebPermissionControllerProxy::messageReceiverName());
 }
 
-void WebPermissionControllerProxy::query(const WebCore::ClientOrigin& clientOrigin, const WebCore::PermissionDescriptor& descriptor, WebPageProxyIdentifier identifier, CompletionHandler<void(std::optional<WebCore::PermissionState>, bool shouldCache)>&& completionHandler)
+void WebPermissionControllerProxy::query(const WebCore::ClientOrigin& clientOrigin, const WebCore::PermissionDescriptor& descriptor, std::optional<WebPageProxyIdentifier> identifier, WebCore::PermissionQuerySource source, CompletionHandler<void(std::optional<WebCore::PermissionState>, bool shouldCache)>&& completionHandler)
 {
-    auto webPageProxy = m_process.webPage(identifier);
+    WebPageProxy* webPageProxy = nullptr;
+    if (source == WebCore::PermissionQuerySource::Window || source == WebCore::PermissionQuerySource::DedicatedWorker) {
+        ASSERT(identifier);
+        webPageProxy = m_process.webPage(identifier.value());
+    } else {
+        std::optional<WeakHashSet<WebProcessProxy>> processes;
+        if (source == WebCore::PermissionQuerySource::SharedWorker)
+            processes = m_process.sharedWorkerClientProcesses();
+        if (source == WebCore::PermissionQuerySource::ServiceWorker)
+            processes = m_process.serviceWorkerClientProcesses();
```

Comment:
This copies the whole WeakHashSet, which seems unneeded.
Maybe we can use a free function to iterate through the hashset without copying it.
The free function will allow to not replicate the code.

Reply:
Yeah I agree, copying the whole WeakHashSet doesn't make sense. I added a function to iterate through but made a private member function so it can access `m_process`.